### PR TITLE
Refactor parser to always return syntax errors

### LINKED ISSUE #1650: Refactor parser from using panics to error return values
### Context

The parser is currently using panics to abort parsing when an unrecoverable parse error is encountered.

### Definition of Done

Replace panics with an explicit error return value

### PR DESCRIPTION
**Pull request by SupunS**
Closes #1650

## Description

Depends on #1718 and #1746

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/parser/expression.go [hunk 1]
```diff
@@ -161,29 +172,36 @@ func defineExpr(def any) {
...
 
 	default:
 		panic(errors.NewUnreachableError())
 	}
+
+	// It is OK to panic here, since this method is only called inside `init()`.
+	// Returning the error from here explodes the error handling in `init` function.
+	// So panic it here, which has the same effect.
+	if err != nil {
+		panic(err)
```
Comment (turbolent) [inline runtime/parser/expression.go:203]: It's already refactored now, but we can keep the panics in the parser "setup" functions, like in `setExprLeftDenotation` – they're only called when the parser package is initialized, not during usage/parsing itself

File: runtime/parser/expression.go [hunk 2]
```diff
@@ -590,7 +632,8 @@ func defineLessThanOrTypeArgumentsExpression() {
 
 			var argumentsStartPos ast.Position
 
-			(func() {
+			// Ignore the error
+			_ = func() error {
```
Comment (turbolent) [inline runtime/parser/expression.go:636]: We can now use the error here to determine if parsing as an invocation succeeded and can remove `isInvocation` now (basically `isInvocation` iff `err != nil`)

File: runtime/parser/transaction.go [hunk 3]
```diff
@@ -52,15 +52,21 @@ func parseTransactionDeclaration(p *parser, docString string) *ast.TransactionDe
 
 	var parameterList *ast.ParameterList
 	if p.current.Is(lexer.TokenParenOpen) {
-		parameterList = parseParameterList(p)
+		parameterList, _ = parseParameterList(p)
```
Comment (turbolent) [inline runtime/parser/transaction.go:55]: Does the error need to be handled here?

File: runtime/parser/expression.go [hunk 4]
```diff
@@ -509,28 +532,47 @@ func init() {
...
 				left,
 				tokenRange.EndPos,
-			)
+			), nil
 		},
 	})
 
 	defineNestedExpression()
-	defineInvocationExpression()
+	err = defineInvocationExpression()
```
Comment (turbolent) [inline runtime/parser/expression.go:545]: As noted above, these setup functions can just keep panicing, they'll never be called in a way where the error needs to be handled specifically

**Review by turbolent — COMMENTED** [review]
Great work!

File: runtime/parser/expression.go (hunk 1 above)
Comment (SupunS) [reply runtime/parser/expression.go]: reverted in 30f9a812d5725ccbbda8a4ca0a4c823738d709af

File: runtime/parser/transaction.go (hunk 3 above)
Comment (SupunS) [reply runtime/parser/transaction.go]: Ah yes, this was automatically added (and hence ignored) by the IDE. Fixed it in 30f9a812d5725ccbbda8a4ca0a4c823738d709af

**Pull request merged**